### ISSUE TITLE
@Binds methods should not be required to have a unique name

### ISSUE THREAD

**Issue by JakeWharton** [google/dagger#1587]
I'm generating an arbitrary number of `@Binds` methods in a single module and I have to contrive unique names for seemingly no good reason. They all semantically perform the same action: binding a generated implementation to a user-written interface. All of these occurrences represent the same thing. There's really no point to a different name as a result, and I've had to resort to just including the type in the method which seems like a waste. It actually makes these `@Binds` methods worse.

These methods will never appear in a stacktrace nor hit by a debugger. They may appear in an error message, but that should include types which disambiguate.

The module is also annotated with `@Generated`, so if you want to conditionally relax it there for binds or all provision methods I'd be happy as well.

**Comment by ronshapiro** [comment]
Having unique names is not just for stack traces or generated factories, but also for error messages.

Even with parameter info, we've often found that users better understand error messages when the method names are unique. Error messages aren't the easiest to understand as it is, so I don't think it's worth removing this check.

An equally fine workaround is to name them `binds<N>` if you don't think that the names are unlikely to be helpful, right?

**Comment by JakeWharton** [comment]
Yeah right now I basically just render the fully-qualified type name with dots replaced with underscores (to avoid simple name collisions). It feels unsatisfying since I'm just duplicating information that is already present in the signature.

At least with provider methods the body might vary which necessitates a meaningful name. For this simple aliasing, it actually seems impossible to provide a unique-yet-semantic name for overloaded binds that isn't duplicating information in the type name(s), duplicating information in the module name, or both.

**Comment by ronshapiro** [comment]
What about using simple names and then deduplicating? Then you at least
have shorter names but decrease the amount of numbers in the method names

בתאריך יום ה׳, 8 באוג׳ 2019, 17:19, מאת Jake Wharton ‏<
notifications@github.com>:

**Comment by JakeWharton** [comment]
Yep, I'll probably make that change. This is less about the right workaround and more about whether or not the check makes sense for binds methods.

If I'm duplicating the type information that should already be displayed in any error message (or other message) which contains the binding does it still make sense to enforce uniqueness?

My argument is that this check actually makes the situation worse because instead of seeing
```
@Binds Foo.Factory toGeneratedFactory(GeneratedFoo_Factory)
```
you now see
```
@Binds Foo.Factory toGeneratedFactoryForFoo(GeneratedFoo_Factory)
```
which is redundant.
